---
title: "R Notebook PS113 Atlantik Microbiome"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook accopanying the manuscript

**High-resolution oceanographic and microbial diversity analyses within trophic groups reveal new provincialism across the Atlantic** . 

When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

Samples were processed using the dada2 pipeline with minor modifications:

primers were trimmed using cutadapt version 1.18
for full details see the scripts and sequencing statistics (Supplementary file gDoc)

## **Data import**

within this script multiple subscripts are combined to make taxonomy, ASV count and metadata operatable and also subsetting for the different trophic level

```{r core_packages, echo=FALSE, message=FALSE, warning=FALSE}
require("tidyverse")
require("ggplot2")
require("dplyr")
```

#### **Primary Productivity and nutrients**

```{r pp, echo=TRUE}
#define oligotrophic and eutrophic provinces
L_CHL <- c("SATL-COLD", "SATL-HOT", "WTRA", "NATR")
H_CHL <- c("FKLD", "BRAZ", "CNRY", "NAST-E", "NADR")
source("C:/Users/choerstm/Documents/core_PS113/PrimaryProductivity/PP_analysis_ext.R", encoding = "UTF-8")

```
#### **Microbial datasets**

sequencing + flow cytometry 

```{r genreal import, echo=FALSE, message=FALSE, warning=FALSE}
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/imports_combined.R", encoding = "UTF-8")
```
## **Data transformations**

#### **clr and hellinger transformations of abundance data**

Sequencing data is due to its technique compositional and needs normalization (Gloor et al. 2017)

*hellinger:*

standard transformation in ecological analyses- gives lower weight to low abundant taxa 
(e.g. https://doi.org/10.1093/femsre/fuw017)
The transformation itself comprises dividing each value in a data matrix by its row sum, and taking the square root of the quotient. (Buttigieg PL, Ramette A (2014) A Guide to Statistical Analysis in Microbial Ecology: a community-focused, living review of multivariate data analyses. FEMS Microbiol Ecol. 90: 543–550.)

*clr:*
The centered log-ratio (clr) transformation has been introduced by Aitchison (1986). Given an observation vector of D “counted” features (taxa, operational taxonomic units or OTUs, genes, etc.) in a sample, x = [x1, x2, ...xD], the clr transformation for the sample can be obtained as follows:
xclr = [log(x1/G(x)), log(x2/G(x)) . . . log(xD/G(x))], 
G(x) = D√x1 · x2 · ... · xD 
G(x) is the geometric mean of x. The clr transformed values can be used as inputs for multivariate hypothesis testing using tools such as MANOVA, regression etc.

```{r, echo=TRUE}
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/general_clr_hellinger_transformation.R")
#hellinger transformation
ASVCount_16S.tax.hellinger <- hellinger(ASVCount_16S.tax,3)
ASVCount_16S.tax.hellinger <- as.data.frame(ASVCount_16S.tax.hellinger)#the 3 is the chosen rowsum cutoff
ASVCount_18S.tax.hellinger <- hellinger(ASVCount_18S.tax,3)
ASVCount_18S.tax.hellinger <- as.data.frame(ASVCount_18S.tax.hellinger)
#clr transformations
ASVCount_16S.tax.clr <- clr(ASVCount_16S.tax,3)#the 3 is the chosen rowsum cutoff
ASVCount_18S.tax.clr <- clr(ASVCount_18S.tax,3)
```

#### **functional subsetting**

rationale for functional groups: trophic grouping based on literature. 
Note that within mixotrophic subsetting might be obligatory heterotrophs as the choosen resoultion is taxonomically not resolved on that level. 

For more information see Table S3

```{r functionalGroups, echo=TRUE}
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/autotrophs_subscript.R", encoding = "UTF-8")
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/mixotrophs_subset.R", encoding = "UTF-8")
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/heterotrophs_subset.R", encoding = "UTF-8")
```

#### **calculate the alpha diversity and bind the estimates to the metadata table** 

use iNEXT package to calculate hill numbers 0-2.

iNEXT focuses on three measures of Hill numbers of order q: species richness (q = 0), Shannon diversity (q = 1, the exponential of Shannon entropy) and Simpson diversity (q = 2, the inverse of Simpson concentration). For each diversity measure, iNEXT uses the observed sample of abundance or incidence data (called the “reference sample”) to compute diversity estimates and the associated 95% confidence intervals for the following two types of rarefaction and extrapolation (R/E):

1. Sample‐size‐based R/E sampling curves: iNEXT computes diversity estimates for rarefied and extrapolated samples up  to an appropriate size. This type of sampling curve plots the diversity estimates with respect to sample size.

2. Coverage‐based R/E sampling curves: iNEXT computes diversity estimates for rarefied and extrapolated samples with sample completeness (as measured by sample coverage) up to an appropriate coverage. This type of sampling curve plots the diversity estimates with respect to sample coverage.

iNEXT also plots the above two types of sampling curves and a sample completeness curve. The sample completeness curve provides a bridge between these two types of curves.

As a good compromise between richness and evenness we proceed with shannon diversity (q=1) in the following analyses


```{r diversity, echo=FALSE}
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/alpha_diversity.R", encoding = "UTF-8")
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/diversity_analyses_combined.R", encoding = "UTF-8")
#diversity_iNEXT_plots(ASVCount_16S.tax) #external calculation and plotting
#diversity_iNEXT_plots(ASVCount_18S.tax) ##external calculation and plotting
meta.16S.div[which.max(meta.16S.div$Shannon),]
meta.16S.div[which.min(meta.16S.div$Shannon),]
meta.18S.div[which.max(meta.18S.div$Shannon),]
meta.18S.div[which.min(meta.18S.div$Shannon),]
```
**subset those samples where we have primary productivity metadata**

```{r sub, echo=TRUE}
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/primary_productivity_subsetting.R", encoding = "UTF-8")
rm(PP_averaged_province)
```

**add the travel calculations to the meta functions**

```{r p_length_scale, echo=TRUE}
pp_data <- c("Station_n", "biomass_transport", "specific_PP", "specific_PP_mol")
meta_functions.PP$Station_n <- as.character(gsub("PS113.S", "", meta_functions.PP$Site))
N_averaged$Station_n <- as.character(N_averaged$Station)
meta_functions.PP <- dplyr::inner_join(meta_functions.PP, N_averaged[pp_data], by="Station_n")
rm(pp_data)
```


```{r, echo=FALSE, message=FALSE}
meta.all.meta_provinces <- meta.all.meta%>%dplyr::group_by(province)%>%
  dplyr::summarise(mean_PC = mean(Total.C..nmol.L., na.rm = TRUE), sd_C = sd(Total.C..nmol.L., na.rm = TRUE), n = dplyr::n(),
            mean_PN = mean(Total.N..nmol.L., na.rm = TRUE), sd_PN = sd(Total.N..nmol.L., na.rm = TRUE))
```


#### **metadata z-scoring**

z scoring needs to be done when metadata with different value distributions are done. in z-scoring the mean of the datavaraible is shifted to 0.

```{r zscore, echo=TRUE}
z_score <- c("salinity", "SST", "PO4_umol.l", "Si_umol.l", "NO2_umol.l", "NO3_umol.l", "oxygen.umol.l.", "Total.N..nmol.L.", "Total.C..nmol.L.", "C.N.Molar", "Distance_coast.km.", "chlorophyll")#, "Richness", "Shannon", "Simpson")
z_score_PP <- c("salinity", "SST", "PO4_umol.l", "Si_umol.l", "NO2_umol.l", "NO3_umol.l", "oxygen.umol.l.", "Total.N..nmol.L.", "Total.C..nmol.L.", "C.N.Molar", "Distance_coast.km.", "PPAVG", "chlorophyll") # "Richness", "Shannon", "Simpson",
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/z-scoring_subset.R")
```

```{r listing, echo=FALSE}
#turn into list to clean up the environment
list_meta <- sapply(ls(pattern="meta_"), function(x) get(x), simplify = FALSE)
rm(list = ls(pattern="meta_"))
list_taxonomy <- sapply(ls(pattern="taxonomy_"), function(x) get(x), simplify = FALSE)
rm(list = ls(pattern="taxonomy_"))
list_ASVCounts <- sapply(ls(pattern="ASVCount_"), function(x) get(x), simplify = FALSE)
rm(list = ls(pattern="ASVCount_"))
list_div <- sapply(ls(pattern="div_"), function(x) get(x), simplify = FALSE)
rm(list = ls(pattern="div_"))
```


```{r hellinger_matrix, echo=FALSE}
#turn hellinger transformations into matrix
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/Hellinger_matrix_tranformation.R")
```


## **Data analysis and plotting**

### primary productivity

```{r pp_figure, echo=FALSE, warning = FALSE, fig.width=7,fig.height=2.5, fig.show="hold", out.width="50%"}
source("C:/Users/choerstm/Documents/core_PS113/PrimaryProductivity/PP_plots.R")
```

### Flow Cytometry
```{r flowplots, echo=FALSE, fig.width=7,fig.height=1.5, fig.show="hold", out.width="50%"}

flow_plots(flow.analysis)
flow.analysis%>%dplyr::filter(province == "CNRY")%>%
  dplyr::summarize(mean(TotalCell.Events.uL., na.rm = TRUE), sd(TotalCell.Events.uL., na.rm = TRUE), dplyr::n())
```
## **Alpha diversity**

#### **province subsetting** 

```{r provinceSubsetting, echo=TRUE}
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/province_subsetting.R", encoding = "UTF-8")
```

### alpha diversity of functions

```{r alpha_div_provinces, echo=FALSE, results='hide', message=FALSE, warning=FALSE, fig.width=5,fig.height=5, fig.show="hold", out.width="50%"}
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/general_corrplot.R", encoding = "UTF-8")
subselection <- c("salinity","SST", "PO4_umol.l","Si_umol.l", "NO3_umol.l", "oxygen.umol.l.", "distance_to_province_boundary.km.",
                  "Total.N..nmol.L.", "Total.C..nmol.L.","chlorophyll","Distance_coast.km.", "Shannon.auto.euk", "Shannon.auto.cynao", "Shannon.mixo",    "Shannon.het.euk", "Shannon.het.prok")
subselection_PP <- c("chlorophyll", "PPAVG", "specific_PP","Shannon.auto.euk", "Shannon.auto.cynao", "Shannon.mixo", "Shannon.het.euk", "Shannon.het.prok", "biomass_transport")
provinces(list_ASVCounts$ASVCount_18S.tax.auto, list_meta$meta_functions.z)
corelation_plot(list_meta$meta_functions.z[subselection], "ALL") #general
corelation_plot(meta_FKLD[subselection], "FKLD")
corelation_plot(meta_BRAZ[subselection], "BRAZ")
corelation_plot(meta_SATL_cold[subselection], "SATL-COLD")
corelation_plot(meta_SATL_hot[subselection], "SATL-HOT")
corelation_plot(meta_WTRA[subselection], "WTRA")
corelation_plot(meta_NATR[subselection], "NATR")
corelation_plot(meta_CNRY[subselection], "CNRY")
corelation_plot(meta_NAST[subselection], "NAST-E")
corelation_plot(meta_NADR[subselection], "NADR")
#PP
provinces(list_ASVCounts$ASVCount_auto_euk.PP, list_meta$meta_functions.PP.z)
corelation_plot(list_meta$meta_functions.PP.z[subselection_PP], "ALL") #general
#corelation_plot(meta_FKLD[subselection_PP])
corelation_plot(meta_BRAZ[subselection_PP], "BRAZ")
corelation_plot(meta_SATL_cold[subselection_PP], "SATL-COLD")
corelation_plot(meta_SATL_hot[subselection_PP], "SATL-HOT")
corelation_plot(meta_WTRA[subselection_PP], "WTRA")
corelation_plot(meta_NATR[subselection_PP], "NATR")
corelation_plot(meta_CNRY[subselection_PP], "CNRY")
corelation_plot(meta_NAST[subselection_PP], "NAST-E")
corelation_plot(meta_NADR[subselection_PP], "NADR")
```

```{r alphadiv, echo=FALSE, fig.width=7,fig.height=3.5}
average_div <- list_meta$meta_functions%>%
        dplyr::group_by(province)%>% 
        dplyr::summarise_at(c("Shannon.auto.euk", "Shannon.auto.cynao", "Shannon.mixo", "Shannon.het.euk", "Shannon.het.prok"), mean)
sd_div <- list_meta$meta_functions%>%
        dplyr::group_by(province)%>% 
        dplyr::summarise_at(c("Shannon.auto.euk", "Shannon.auto.cynao", "Shannon.mixo", "Shannon.het.euk", "Shannon.het.prok"), sd)
av_div_l <- gather(average_div, trophy, mean, Shannon.auto.euk, Shannon.auto.cynao, Shannon.mixo, Shannon.het.euk, Shannon.het.prok)
sd_div_l <- gather(sd_div, trophy, sd, Shannon.auto.euk, Shannon.auto.cynao, Shannon.mixo, Shannon.het.euk, Shannon.het.prok)
av_tot <- dplyr::inner_join(av_div_l, sd_div_l, by = c("province", "trophy"))
#remove the sample without proper province
discard <- apply(av_tot, 1, function(x) any(is.na(x)))
av_total <- av_tot[!discard,]
p<- ggplot(av_total, aes(x=province, y=mean, fill=trophy)) + 
  geom_bar(stat="identity", color="black", 
           position=position_dodge()) +
  geom_errorbar(aes(ymin=mean-sd, ymax=mean+sd), width=.2,
                 position=position_dodge(.9)) 
p+labs(title="Shannon diversity", x="province", y = "Shannon entropy hill = 1")+
   theme_classic() +
   scale_fill_manual(values=c('#72E815','#2E7220', '#5B3E1E', '#B27C42', '#D6BF16'))

```


```{r correlation_plots, echo=FALSE, fig.width=7,fig.height=3.5, fig.show="hold", out.width="50%"}
#Individual correlation analyses as examples of correlations between functional groups and environmental parameters, e.g. temperature
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/correlation_individual.R")
```
**wilcoxon test**

```{r wilcoxon_mixo, echo=TRUE, warning=FALSE}
require("ggpubr")
L_CHL_mixo <- list_meta$meta_mixo%>%dplyr::filter(province %in% L_CHL)
H_CHL_mixo <- list_meta$meta_mixo%>%dplyr::filter(province %in% H_CHL)

print(wilcox.test(L_CHL_mixo$Shannon, H_CHL_mixo$Shannon,  var.equal = TRUE))
```


#### residuals

```{r cor_residulas, echo = FALSE, message = FALSE, fig.show="hold", out.width="25%", fig.width=3.5,fig.height=3.5}
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/Residuals_corrplot.R")
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/Cor_residuals_correlations.R")
#clean up!
rm(list = ls(pattern="meta_"))
rm(list = ls(pattern="abundance_"))
rm(list = ls(pattern="abundance."))
rm(a.residuals)
```

#### PCA and PCoA of environmental metadata

- look at the eigenvalues, if they are negative then they don't make sense and shouldn't be included in further analysis
the problem with PCoA is that it sorts the eigenvalues and you don't know which is what

```{r metadata_analysis, echo=TRUE, warning=FALSE, fig.width=8,fig.height=6, fig.show="hold", out.width="50%"}
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/general_metadata_analysis.R", encoding = "UTF-8")
#define the import data
meta_sub <- c("Site","province", "SST", "salinity", "distance_to_province_boundary.km.", "Distance_coast.km.", "oxygen.umol.l.","NO3_umol.l","PO4_umol.l","Total.N..nmol.L.", "Total.C..nmol.L.", "chlorophyll")
meta_sub_PP <- c("Site","province", "SST", "salinity", "distance_to_province_boundary.km.", "Distance_coast.km.", "oxygen.umol.l.","NO3_umol.l","PO4_umol.l","Total.N..nmol.L.", "Total.C..nmol.L.", "chlorophyll", "PPAVG", "specific_PP", "biomass_transport")
meta_PCA(list_meta$meta_functions.z[meta_sub])
meta_PCA(list_meta$meta_functions.PP.z[meta_sub_PP])
me_data <- c("SST", "salinity", "distance_to_province_boundary.km.", "Distance_coast.km.", "oxygen.umol.l.","NO3_umol.l","PO4_umol.l","Total.N..nmol.L.", "Total.C..nmol.L.", "chlorophyll")
me_data_PP <- c("SST", "salinity", "distance_to_province_boundary.km.", "Distance_coast.km.", "oxygen.umol.l.","NO3_umol.l","PO4_umol.l","Total.N..nmol.L.", "Total.C..nmol.L.", "chlorophyll", "PPAVG")
meta_PCoA(list_meta$meta_functions.z[me_data],list_meta$meta_functions.z["province"], list_meta$meta_functions.z["Site"])
meta_PCoA(list_meta$meta_functions.PP.z[me_data_PP],list_meta$meta_functions.PP.z["province"], list_meta$meta_functions.PP.z["Site"])
rm(meta_sub, meta_sub_PP, me_data, me_data_PP)
```

### **RDA of functions**
```{r RDA, echo=FALSE, fig.width=6,fig.height=5, fig.show="hold", out.width="50%"}
source("C:/Users/choerstm/Documents/core_PS113/Genomics/RScripts/RDAs_functions.R", encoding = "UTF-8")
#prepare data for RDA
RDA_meta <- c("Site","province", "geo_current", "SST", "salinity", "distance_to_province_boundary.km.", "Distance_coast.km.", "oxygen.umol.l.","NO3_umol.l","PO4_umol.l","Total.N..nmol.L.", "Total.C..nmol.L.", "chlorophyll")
RDA_meta_PP <- c("Site","province", "geo_current", "SST", "salinity", "distance_to_province_boundary.km.", "Distance_coast.km.", "oxygen.umol.l.","NO3_umol.l","PO4_umol.l","Total.N..nmol.L.", "Total.C..nmol.L.", "chlorophyll", "PPAVG")
#all
RDA_plots(list_meta$meta_auto_euk.z[c(RDA_meta)], list_ASVCounts$ASVCount_18S.tax.auto.clr, list_ASVCounts$ASVCount_18S.tax.auto.hellinger)
RDA_plots(list_meta$meta_auto_cyano.z[c(RDA_meta)], list_ASVCounts$ASVCount_16S.tax.auto.clr, list_ASVCounts$ASVCount_16S.tax.auto.hellinger)
RDA_plots(list_meta$meta_mixo.z[c(RDA_meta)], list_ASVCounts$ASVCount_mixo.clr, list_ASVCounts$ASVCount_mixo.hellinger)
RDA_plots(list_meta$meta_het_euk.z[c(RDA_meta)], list_ASVCounts$ASVCount_18S.tax.het.clr, list_ASVCounts$ASVCount_18S.tax.het.hellinger)
RDA_plots(list_meta$meta_het_prok.z[c(RDA_meta)], list_ASVCounts$ASVCount_16S.tax.het.clr, list_ASVCounts$ASVCount_16S.tax.het.hellinger)
#PP
RDA_plots(list_meta$meta_auto_euk.PP.z[c(RDA_meta_PP)], list_ASVCounts$ASVCount_auto_euk.PP.clr, list_ASVCounts$ASVCount_auto_euk.PP.hellinger)
RDA_plots(list_meta$meta_auto_cyano.PP.z[c(RDA_meta_PP)], list_ASVCounts$ASVCount_auto_cyano.PP.clr, list_ASVCounts$ASVCount_auto_cyano.PP.hellinger)
RDA_plots(list_meta$meta_mixo.PP.z[c(RDA_meta_PP)], list_ASVCounts$ASVCount_mixo.PP.clr, list_ASVCounts$ASVCount_mixo.PP.hellinger)
RDA_plots(list_meta$meta_het_euk.PP.z[c(RDA_meta_PP)], list_ASVCounts$ASVCount_het_euk.PP.clr, list_ASVCounts$ASVCount_het_euk.PP.hellinger)
RDA_plots(list_meta$meta_het_prok.PP.z[c(RDA_meta_PP)], list_ASVCounts$ASVCount_het_prok.PP.clr, list_ASVCounts$ASVCount_het_prok.PP.hellinger)
```

